docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,277 @@
+    is updated in status based on latest observation.
+    
+    “defaultFieldB” if not provided by K8s user, gets updated to “valueB” by reconciler.
+    
+    overrideFieldC‘s is modified by reconciler whenever provided by the k8s user. Ex: “-xyz” gets appended to it’s value. 
+
+    ```yaml
+    Spec:
+     fieldA string
+     defaultFieldB string
+     overrideFieldC string
+    Status:
+     fieldD string
+    ```
+
+    and the manifest file(desired state) looks like,
+
+    ```yaml
+    Spec:
+     fieldA: valueA
+     overrideFieldC: valueC
+   ```
+
+2.  When this resource is reconciled by the ACK service controller, we expect the CreateOutput to have following value 
+    for these fields. fieldA → “valueA“, defaultFieldB → “valueB“, overrideFieldC → “valueC-xyz“, fieldD → ”valueD“
+    
+    Based on this result, ACK service controller will persist the following object in etcd:
+    
+    Annotations:
+    
+    ```yaml
+    ack/spec-overrides: 
+       “{
+           \“defaultFieldB\”:
+               {
+                   \“serverOverride\”:\“valueB\”
+               }, 
+           \“overrideFieldC\”:
+               {
+                   \“userInput\”:\“valueC\”,
+                   \“serverOverride\”:\“valueC-xyz\”
+               }
+       }”
+       Spec:
+        fieldA: valueA
+        defaultFieldB: valueB
+        overrideFieldC: valueC-xyz
+       Status:
+        fieldD: valueD
+    ```
+   
+    > Note: the annotation above has single string value. It is been broken into multiple lines to showcase the structure better.
+   
+    > Note: Fields whose value are defaulted, do not contain “userInput” and that is how they are distinguished.
+   
+    > Note: The change will be made in the sdkCreate and sdkUpdate methods, in the code-generator repository. Whenever there is an update required for spec field, based on the Create/Update output shape, we will record it in the Annotations, so that during next reconciliation, this context is preserved.
+   
+    > Note: sdkCreate will marshall the spec field and use string comparison to determine whether spec needs update or not. This will prevent from runtime panic of incomparable fields while using “==” or DeepEqual() method. 
+
+3. During the next reconciliation of manifest mentioned in #1, the desired state will be constructed not just from spec 
+   in etcd but also the Annotations. and the desired object will look like the object persisted in etcd in #2
+
+   After reading the desired state from etcd, the reconciler code will go through all the spec fields and see whether 
+   they exist in the spec-override annotation map. If a corresponding entry exists, reconciler will marshall current 
+   desired value of spec field and compare with “userInput” in annotation override, 
+
+   * If there is no difference between “userInput” and marshalled desiredValue, reconciler will update the desired state
+     with serverOverride from spec-override annotation. 
```

Comment:
Just throwing this crazy idea out there - I hope that no AWS service has implemented this:
What if the server side value is non-deterministic? That is, we give it the same value on the `update` but it returns a different `serverOverride`?

Reply:
@RedbackThomson ElastiCache's `User.AccessString` field is an example of this. The value that the k8s user places in the `User.AccessString` field is *always* overridden by the ElastiCache service and returned as the `AccessString` in the `DescribeUser` API call.

Reply:
@RedbackThomson I'm not sure that it's nondeterministic – although we can't necessarily translate between the user-specified `AccessString` and the server-returned `AccessString`, you should get the same thing every time. However, we should probably assume that the way the expansion/translation is done on the server-side is subject to change.

more info here: https://github.com/aws-controllers-k8s/elasticache-controller/pull/32